util/app/discovery/discovery.go

```diff
@@ -99,109 +90,167 @@ func DetectConfigManagementPlugin(ctx context.Context, appPath, repoPath, plugin
+	NewConfigManagementPluginClient() (pluginclient.ConfigManagementPluginServiceClient, io.Closer, error)
+}
+
+// socketCMPClientConstructor is a CMPClientConstructor for a CMP that uses a socket file.
+type socketCMPClientConstructor struct {
+	pluginSockFilePath string
+	filename           string
+}
+
+// newSocketCMPClientConstructorForPath returns a new socketCMPClientConstructor.
+func newSocketCMPClientConstructorForPath(pluginSockFilePath, filename string) CMPClientConstructor {
+	return socketCMPClientConstructor{
+		pluginSockFilePath: pluginSockFilePath,
+		filename:           filename,
+	}
+}
+
+// newSocketCMPClientConstructorForPluginName returns a new socketCMPClientConstructor for the given plugin name.
+func newSocketCMPClientConstructorForPluginName(pluginSockFilePath, pluginName string) CMPClientConstructor {
+	return newSocketCMPClientConstructorForPath(pluginSockFilePath, pluginName+".sock")
+}
+
+// NewConfigManagementPluginClient returns a client for the CMP. It also returns a closer for the connection.
+func (c socketCMPClientConstructor) NewConfigManagementPluginClient() (pluginclient.ConfigManagementPluginServiceClient, io.Closer, error) {
+	absPluginSockFilePath, err := filepath.Abs(c.pluginSockFilePath)
 	if err != nil {
-		log.Errorf("error getting absolute path for plugin socket dir %v, %v", pluginSockFilePath, err)
-		return nil, nil, false
+		return nil, nil, fmt.Errorf("error getting absolute path for plugin socket dir %q: %w", c.pluginSockFilePath, err)
 	}
-	address := filepath.Join(absPluginSockFilePath, fileName)
-	if !files.Inbound(address, absPluginSockFilePath) {
-		log.Errorf("invalid socket file path, %v is outside plugin socket dir %v", fileName, pluginSockFilePath)
-		return nil, nil, false
+	address, err := securejoin.SecureJoin(absPluginSockFilePath, c.filename)
+	if err != nil {
+		return nil, nil, fmt.Errorf("invalid socket file path, %v is outside plugin socket dir %v: %w", c.filename, c.pluginSockFilePath, err)
 	}
-
 	cmpclientset := pluginclient.NewConfigManagementPluginClientSet(address)
-
 	conn, cmpClient, err := cmpclientset.NewConfigManagementPluginClient()
 	if err != nil {
-		log.WithFields(log.Fields{
-			common.SecurityField:    common.SecurityMedium,
-			common.SecurityCWEField: common.SecurityCWEMissingReleaseOfFileDescriptor,
-		}).Errorf("error dialing to cmp-server for plugin %s, %v", fileName, err)
-		return nil, nil, false
+		return nil, nil, fmt.Errorf("error dialing to cmp-server for plugin: %w", err)
 	}
+	return cmpClient, conn, nil
+}
 
-	cfg, err := cmpClient.CheckPluginConfiguration(ctx, &empty.Empty{})
+// getClientForFilepath returns a client for the given filepath. It also returns a closer for the connection and
+// a boolean indicating if the plugin has discovery configured.
+func getClientForFilepath(ctx context.Context, c CMPClientConstructor) (pluginclient.ConfigManagementPluginServiceClient, io.Closer, bool, error) {
```

Comment:
Renamed to `getClientAndConfig`. The purpose of the function shifted a bit as I refactored. Now it basically just constructs the CMP client and does the initial call to get the CMP's config. It's in a function because it's called by both `namedCMPSupports` and `unnamedCMPSupports`. It's only 8 lines, so I could duplicate if that feels more reasonable.